File: Ecclesia/app/Models/Campaign.php
<?php

declare(strict_types=1);

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Campaign extends Model
{
    protected $fillable = [
        'parish_id',
        'title',
        'description',
        'type',
        'target_amount',
        'collected_amount',
        'image',
        'is_active',
        'ends_at',
    ];

    /**
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'target_amount' => 'integer',
            'collected_amount' => 'integer',
            'is_active' => 'boolean',
            'ends_at' => 'date',
        ];
    }

    /**
     * @return BelongsTo<Parish, $this>
     */
    public function parish(): BelongsTo
    {
        return $this->belongsTo(Parish::class);
    }

    /**
     * @return HasMany<CampaignPledge, $this>
     */
    public function pledges(): HasMany
    {
        return $this->hasMany(CampaignPledge::class);
    }

    public function typeLabel(): string
    {
        return match ($this->type) {
            'cotisation' => 'Cotisation',
            'quete' => 'Quête',
            default => 'Don',
        };
    }

    public function progress(): int
    {
        if (! $this->target_amount || $this->target_amount <= 0) {
            return 0;
        }

        return (int) min(100, round($this->collected_amount / $this->target_amount * 100));
    }
}

File: ecclesia_/lib/features/donations/data/campaign.dart
class Campaign {
  const Campaign({
    required this.id,
    required this.title,
    required this.typeLabel,
    this.description,
    this.targetAmount,
    this.collectedAmount = 0,
    this.progress = 0,
    this.imageUrl,
    this.endsAt,
  });

  final int id;
  final String title;
  final String typeLabel;
  final String? description;
  final int? targetAmount;
  final int collectedAmount;
  final int progress;
  final String? imageUrl;
  final String? endsAt;

  factory Campaign.fromJson(Map<String, dynamic> json) => Campaign(
        id: json['id'] as int,
        title: json['title'] as String? ?? '',
        typeLabel: json['type_label'] as String? ?? 'Don',
        description: json['description'] as String?,
        targetAmount: (json['target_amount'] as num?)?.toInt(),
        collectedAmount: (json['collected_amount'] as num?)?.toInt() ?? 0,
        progress: (json['progress'] as num?)?.toInt() ?? 0,
        imageUrl: json['image_url'] as String?,
        endsAt: json['ends_at'] as String?,
      );
}
